### PR TITLE
Add AppleApp to Firebase.

### PR DESCRIPTION
**Pull request by mraouffouad**
Add AppleApp resource to Firebase.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource

```

### CODE REVIEW

File: mmv1/products/firebase/api.yaml
Comment (tylerg-dev) [inline mmv1/products/firebase/api.yaml:167]: probably use https://firebase.google.com/docs/ios/setup instead (or both) as it walks through the Apple ecosystem parts which are more directly applicable then the general Firebase docs landing page.

File: mmv1/products/firebase/terraform.yaml
Comment (tylerg-dev) [inline mmv1/products/firebase/terraform.yaml:61]: I think this should be:
```
'projects/{{project}}/iosApps/{{appId}}', 'iosApps/{{appId}}', '{{appId}}'] 
```

Since the resource's AppId is a GUID, we can support all 3 variants without ambiguity.

File: mmv1/products/firebase/terraform.yaml
Comment (tylerg-dev) [inline mmv1/products/firebase/terraform.yaml:63]: We support Delete on AppleApp, so lets not do this.

we should not pull the new resource in crippled.  Slower updated to WebApp are not as big an issue as it already has that gap.

File: mmv1/templates/terraform/examples/firebase_apple_app_basic.tf.erb
Comment (tylerg-dev) [inline mmv1/templates/terraform/examples/firebase_apple_app_basic.tf.erb:8]: we probably don't need to re-create the Project each time, if the App is being deleted (as per above comments)

File: mmv1/templates/terraform/examples/firebase_apple_app_basic.tf.erb
Comment (tylerg-dev) [inline mmv1/templates/terraform/examples/firebase_apple_app_basic.tf.erb:14]: I think we likely want to have 2 samples in general:
- a minimal sample that sets only the minimally required fields
- a full sample that sets all optional values with example values

File: mmv1/products/firebase/terraform.yaml
Comment (tylerg-dev) [inline mmv1/products/firebase/terraform.yaml:74]: we might be better off using test_env_overrides for the IDs to use semi-realistic values
ex:   
```bundle_id: "apple.app." + randString(t, 5)" ```

File: mmv1/templates/terraform/examples/firebase_apple_app_basic.tf.erb
Comment (tylerg-dev) [inline mmv1/templates/terraform/examples/firebase_apple_app_basic.tf.erb:17]: does this pass tests?  bundle-id is non-optional....

File: mmv1/templates/terraform/examples/firebase_apple_app_basic.tf.erb
Comment (tylerg-dev) [inline mmv1/templates/terraform/examples/firebase_apple_app_basic.tf.erb:16]: nit: I'd put ```google_firebase_project.default.project ``` here instead.  that makes it abundantly clear the relationshiop.

File: mmv1/products/firebase/terraform.yaml
Comment (tylerg-dev) [inline mmv1/products/firebase/terraform.yaml:77]: does something like 'randInt(t, 10)' work here?  not sure which part of the ecosystem is resolving the random numbers...

**Review by megan07 — DISMISSED** [review]
Hi @mraouffouad, sorry for the delay in review, it looks like there is an error with delete, like the `appId` wasn't showing up, I'm hoping switching to `name` (which would be a value from state that includes that id) will help.

Thanks!

File: mmv1/products/firebase/api.yaml
Comment (rainshen49) [inline mmv1/products/firebase/api.yaml:193]: Prefer same wording as WebApp, just different default
```
(Optional) Set to `ABANDON` to allow the AppleApp to be untracked from terraform state
          rather than deleted upon `terraform destroy`. This is useful because the AppleApp may be 
          serving traffic. Set to `DELETE` to delete the AppleApp. Default to `DELETE`
```

File: mmv1/templates/terraform/custom_delete/firebase_app_deletion_policy.erb
Comment (rainshen49) [inline mmv1/templates/terraform/custom_delete/firebase_app_deletion_policy.erb:1]: If you rebase, this file should be gone from the PR

**Comment by mraouffouad** [comment]
> Hi @mraouffouad, sorry for the delay in review, it looks like there is an error with delete, like the `appId` wasn't showing up, I'm hoping switching to `name` (which would be a value from state that includes that id) will help.
> 
> Thanks!

Fixed, Thanks.

**Comment by mraouffouad** [comment]
@megan07 This PR is blocking our further Terraform tasks and is becoming urgent. Please take a look. FWIW, the parallel resource AndroidApp has just been merged: https://github.com/GoogleCloudPlatform/magic-modules/pull/6629

**Comment by mraouffouad** [comment]
@melinath This PR is ready for review, and is currently blocking our further work. Please take a look.

**Review by melinath — CHANGES_REQUESTED** [review]
First set of change requests - thanks!

File: mmv1/products/firebase/api.yaml
Comment (tylerg-dev) [reply mmv1/products/firebase/api.yaml:226]: Firebase Public Docs refer to Apple platforms etc. which is what this entity represents.  The API's naming is from 2018 and will be changed in the future when a GA version is released.

When we were using DCL->Terraform as our path, this naming was approved after discussion.

Comment (melinath) [reply mmv1/products/firebase/api.yaml:226]: 👍 sounds good.

**Review by melinath — CHANGES_REQUESTED** [review]
In addition to the remaining changes requested above, please add an additional handwritten [update test](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/README.md#update-tests)

**Comment by mraouffouad** [comment]
> In addition to the remaining changes requested above, please add an additional handwritten [update test](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/README.md#update-tests)

I added an update test. I believe all other changes are addressed. Please let me know if I missed anything.

**Review by melinath — CHANGES_REQUESTED** [review]
see unresolved comments above

**Comment by mraouffouad** [comment]
> see unresolved comments above

I'm not sure if these need more actions. I left them open in case you want to comment further, otherwise I will go ahead and resolve.

File: mmv1/products/firebase/terraform.yaml
Comment (melinath) [reply mmv1/products/firebase/terraform.yaml:99]: @mraouffouad please action this item.

File: mmv1/products/firebase/terraform.yaml
Comment (melinath) [reply mmv1/products/firebase/terraform.yaml:118]: @mraouffouad please respond to this question and remove the randomness if it's not necessary.

File: mmv1/products/firebase/terraform.yaml
Comment (melinath) [reply mmv1/products/firebase/terraform.yaml:99]: The docs currently generated show

```
  bundle_id = ""
  app_store_id = ""
  team_id = ""
```

Which is not useful as an example.

https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6630-old..auto-pr-6630#diff-acc4a59b9f4bc11e27c525abbc43c1e21768246855f386c057f2f0f4d6ade0bbR53

File: mmv1/products/firebase/terraform.yaml
Comment (melinath) [reply mmv1/products/firebase/terraform.yaml:118]: If the randomness is not necessary, these lines could even be removed and you could just rely on the primary setting of these variables in `vars` (once that's added.)

File: mmv1/third_party/terraform/tests/resource_firebase_apple_app_update_test.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_firebase_apple_app_update_test.go.erb:15]:
```suggestion
                "bundle_id":  "apple.app.12345",
```

File: mmv1/third_party/terraform/tests/resource_firebase_apple_app_update_test.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_firebase_apple_app_update_test.go.erb:24]:
```suggestion
				Config: testAccFirebaseAppleApp(context, 12345, "1"),
```

File: mmv1/third_party/terraform/tests/resource_firebase_apple_app_update_test.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_firebase_apple_app_update_test.go.erb:27]:
```suggestion
				Config: testAccFirebaseAppleApp(context, 67890, "2"),
```

File: mmv1/products/firebase/terraform.yaml
Comment (melinath) [inline mmv1/products/firebase/terraform.yaml:111]:
```suggestion
          bundle_id: "apple.app.12345"
```

File: mmv1/products/firebase/terraform.yaml
Comment (melinath) [inline mmv1/products/firebase/terraform.yaml:101]:
```suggestion
          bundle_id: "apple.app.12345"
```

File: mmv1/products/firebase/terraform.yaml
Comment (melinath) [inline mmv1/products/firebase/terraform.yaml:117]: These overrides are no longer required since these vars are the same as in `vars:`

```suggestion
```

**Review by melinath — CHANGES_REQUESTED** [review]
looking really good! Just a few minor cleanups noted below

**Comment by mraouffouad** [comment]
> looking really good! Just a few minor cleanups noted below

Thanks Stephen for the suggestions. All is done. My only concern is that I'm not sure if removing the test_vars_overrides section will pass the tests. As I mentioned before team_idand app_store_id are int and moving them to the vars section will append their set values by random string in the test, and therefore fail our API validators. I'm willing to give it one more try and let's see how the tests go.

**Comment by mraouffouad** [comment]
> > looking really good! Just a few minor cleanups noted below
> 
> Thanks Stephen for the suggestions. All is done. My only concern is that I'm not sure if removing the test_vars_overrides section will pass the tests. As I mentioned before team_id and app_store_id are integers and moving them to the vars section will append their set values by random string in the test, and therefore fail our API validators. I tried this a few times before and always got "Error: Error creating AppleApp: googleapi: Error 400: Request contains an invalid argument". I decided to give it one more try and let's see how the tests go.

As expected, test failed because of the added random suffixes to these 2 variables.

{
 "appStoreId": "**12345rnoosq6yng**",
 "bundleId": "apple.app.12345rnoosq6yng",
 "displayName": "Display Name Fullrnoosq6yng",
 "teamId": "**9987654321rnoosq6yng**"
}

-----------------------------------------------------
2022/11/11 21:06:06 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 400 Bad Request
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Fri, 11 Nov 2022 21:06:06 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "error": {
    "code": 400,
    "message": "Request contains an invalid argument.",
    "status": "INVALID_ARGUMENT"
  }
}

**Comment by mraouffouad** [comment]
@melinath is there anything blocking this PR at this point?

File: mmv1/products/firebase/terraform.yaml
Comment (melinath) [reply mmv1/products/firebase/terraform.yaml:117]: per discussion - these are required to avoid a random suffix on these integer fields.

**Comment by melinath** [comment]
LGTM - doing a manual test run just to be safe: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/353024

**Review by melinath — APPROVED** [review]

**Pull request merged**
